### PR TITLE
[SPARK-18838][CORE] Introduce multiple queues in LiveListenerBus

### PR DESCRIPTION
**Pull request by bOOm-X**
## What changes were proposed in this pull request?

In this PR the single queue of the LiveListenerBus was replaced by multiple independent queues.
 - The EventLoggingListener was put in an independent queue (it was the most time consuming listener).
 - The executorAllocationManager was put in an independent qeue (it is very impacted by event drop of the main queue due to other listener impacts, so isolation purpose)
 - The UI listeners were put in the same independent queue (One of them was very time consuming, we want no impact of the GUI on the other listener)
 - The "extralisteners" were put in the same independent queue, for isolation purpose
 - The streamingListenerBus was put in an independent queue. this llistener is a bus too, and the processing time of its listeners can be quite significant

The queue and its processing thread have been extracted from `LiveListenerBus` into a class `BusQueue`. The definition of most of the methods of `ListenerBus` has been extracted into a trait `WithListenerBus`. The `LiveListenerBus` implements it directly. It hold  the "default" queue associated with a group of listeners and a list of queues. The method `addListener`  of the `WithListenerBus` has a new optional boolean parameter (default value false) to require a independent queue for this listener instead of the default one. This parameter is ignored in the default implementation (in`ListenerBus`) . 
A listener which is also a set of listeners has been added. It allows to keep the current behavior for a group of dependent listeners or the default queue. It handles the per listener metrics.
The methods `addProcessor` and `removeProcessor` have been added to `LiveListenerBus` to be able to add message processing at the super type `SparkListenerEvent` in addition to the per event type processing of the listener interface. 

## How was this patch tested?
utest + manual tests on the cluster

### CODE REVIEW

**Comment by bOOm-X** [comment]
@vanzin @cloud-fan Can I have a review on this new PR ?

**Comment by vanzin** [comment]
I'm busy, but I'll get to it eventually. You could at least write a proper commit summary in the meantime.

**Comment by vanzin** [comment]
ok to test

**Comment by bOOm-X** [comment]
Ok I rebased & updated my commit message.
I will update the PR message and push more commit to use what is done.

**Comment by vanzin** [comment]
I took a quick look and this does indeed look very much like work in progress. I also have a feeling that it's way over-engineered; there's a lot of base classes that are not that interesting, for example:

- why `SynchronousListenerBus`? It doesn't seem to add anything interesting over the existing interface.
- the whole "group of listeners" abstraction seems unnecessary. As far as I see it can be folded into the "listener queue" concept -  a queue feeds events to a collection of listeners.

Changing "post" to "postToAll" as part of this change also is adding a lot of unnecessary noise. I'm not a fan of the current class hierarchy of the listener bus and I think that change makes sense, but at the same time it should be done separately since it's distracting here.

I also saw methods that are not fully implemented in the code, so I assume you're still working on this.

I'd also like to see better justification for your custom queue implementation. Have you identified the use of `BlockingQueue` as a hotspot in the current code? My main worry is the sleep; if you're unlucky, your queue will always be 20ms behind in processing events and may suffer if there's a sudden burst while it's sleeping. You can probably squeeze better performance out of `BlockingQueue` without having to write your own - e.g. by using `drainTo` instead of reading events one by one.

The approach in #16291 had a lot of good things going for it, and mostly needed some clean up (and be modified to only change the live listener bus, and not the replay one). Your current approach seems a lot more complicated than that.

**Comment by bOOm-X** [comment]
@vanzin Ok it is ready now.

> why SynchronousListenerBus? It doesn't seem to add anything interesting over the existing interface.

I removed it.

> the whole "group of listeners" abstraction seems unnecessary. As far as I see it can be folded into the "listener queue" concept - a queue feeds events to a collection of listeners.

I simplified it and add usage in the other commits. It is basically usefull to hold the metrics, and I need a common way to add a group of dependent listeners to the LiveListenerBus and the ReplayBus.

> Changing "post" to "postToAll" as part of this change also is adding a lot of unnecessary noise. I'm not a fan of the current class hierarchy of the listener bus and I think that change makes sense, but at the same time it should be done separately since it's distracting here

100% agree. I removed it

> I'd also like to see better justification for your custom queue implementation. Have you identified the use of BlockingQueue as a hotspot in the current code?

This implementation had 2 advantages: it is a 1 producer - 1 consumer queue whereas the BlockingQueue is a n producers - m consumers. So it use much less synchronization. The other advantage (the main one) is that no object is created for each message added to the queue. So it produces a lot less garbage. More independent queues we have, more it is significant.

> My main worry is the sleep; if you're unlucky, your queue will always be 20ms behind in processing events and may suffer if there's a sudden burst while it's sleeping. 

I change it to 1 ms instead of 20ms. This time is much less than the average processing time of the fastest listener (around 5 ms for the  HeartbeatListener). It is just to force the consumer thread to escape in case of empty queue to give more chance to the producer thread to be scheduled. I can remove it if you want.

**Comment by vanzin** [comment]
> This implementation had 2 advantages... it use much less synchronization.

Yes, but have you quantified how much you win with that? If the blocking queue approach has enough throughput for the listener bus, it's safer to use it.

> is that no object is created for each message added to the queue

Well, you could use an `ArrayBlockingQueue`. Then no extra object is allocated either.

Here's a link with numbers for `ArrayBlockingQueue`:
https://github.com/LMAX-Exchange/disruptor/wiki/Performance-Results

4M ops per sec in the 1P-1C case looks plenty fast for Spark's need.

> I change it to 1 ms instead of 20ms. 

I think if you really insist on going this route, you should use `LockSupport.park/unpark` instead of fighting for CPU time like this.

**Comment by bOOm-X** [comment]
> Well, you could use an ArrayBlockingQueue. Then no extra object is allocated either.

Yes I agree. But you get the synchronization too. I am still agree that it should not have a big impact yet. But using an ArrayBlockingQueue does not simplify the code a lot. The current implementation is not complicated, not too verbose, and base on a simple pure scala array. I do not think that it has a huge complexity cost compared to the java ArrayBlockingQueue.

I change the Thread.sleep to a Thread.yield to be less agressive for the thread unscheduling. Even with very few messages it should not consume too much CPU, and it will be much more reactive when messages are bursting.

**Comment by vanzin** [comment]
> I change the Thread.sleep to a Thread.yield to be less agressive for the thread unscheduling

How does that help? Now you have a thread that never sleeps. It gives up its scheduler slot but is still active and will be scheduled again, even if it doesn't need to.

Can you explain in numbers why exactly you're against any of the other solutions that actually prevent these threads from being scheduled when not needed?

**Comment by vanzin** [comment]
To illustrate what I'm trying to say, here's what `top` says for an *idle* spark-shell with your code:

```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                             
24247 vanzin    20   0 6778304 766336  31836 S 201.0  2.3   1:34.06 java  
```

I hope you see why that's an issue.

**Comment by bOOm-X** [comment]
With a `LockSupport.parkNanos(1000)` instead I get:

```
PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
43448 a.prang   20   0 17.663g 2.378g  66668 S  35.2  0.9   3:41.04 java
```
Is it ok ?

**Comment by vanzin** [comment]
And idle spark-shell should be using 0% of the CPU.

```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                             
 1304 vanzin    20   0 6773996 743104  31468 S   0.3  2.3   0:37.28 java
```

**Comment by bOOm-X** [comment]
please test

**Comment by bOOm-X** [comment]
`PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
11312 a.prang   20   0 17.661g 1.699g  66924 S   8.3  0.7   1:53.41 java`

Ok it is good now.
Do you have an idea why the PySpark packaging test are failling ?
I tried to rebase but no impact

**Comment by bOOm-X** [comment]
@vanzin : Any news ?

**Comment by cloud-fan** [comment]
is this a successor of https://github.com/apache/spark/pull/16291 ?

**Comment by bOOm-X** [comment]
@cloud-fan Yes it is !

**Comment by bOOm-X** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
What's the processing model now? Previously it's very simple, we have an event queue, and many threads send events to this queue, and one thread poll events from this and call listeners. What is it now?

Can you give a summary about the implementation? e.g. How you change the API in `ListenerBus` and what's the new class hierarchy for listener buses?

**Comment by vanzin** [comment]
I was out for the last couple of weeks. But, taking a quick look, I really would prefer for you to stop trying to implement your own blocking queue without numbers that justify the need for it. Putting event listeners in separate queues does not require you to write a custom blocking queue implementation; if there's benefit in that, you can do it as a separate PR, when you actually provide evidence that backs up the necessity for it.

**Comment by bOOm-X** [comment]
@vanzin I put an ArrayBlockingQueue as you wanted

**Comment by bOOm-X** [comment]
@cloud-fan 
There are now multiple queues. Many threads send events to a dispatcher which copy the event in each queue. One thread per queue poll the events from its dedicated queue and call its attached listener(s).

For listener bus there is now two pure traits. the first one WithListenerBus allow to add / remove listener (current behavior). By default all the current implementation implement this trait . Another trait WithMultipleListenerBus inherit from the previous one and add a method to add isolated listener. Only the LiveListenerBus and the ReplayListenerBus implement this trait. For the second one, it is just a synonym of the simple addListener method. All the UI listeners are added to both through this method as a group of listeners (they are inter-dependent).
The StreamingListenerBus, the EventLoggingListener, the ExecutorAllocationManager are added to the LiveListenerBus (with this method) to a dedicated queue. The "extralisteners" are added to it as a group of listeners in a dedicated queue too. All other listeners are added to the "default" queue with the old method.

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 1]
```diff
@@ -532,7 +533,10 @@ class SparkContext(config: SparkConf) extends Logging {
           new EventLoggingListener(_applicationId, _applicationAttemptId, _eventLogDir.get,
             _conf, _hadoopConfiguration)
         logger.start()
-        listenerBus.addListener(logger)
+        listenerBus.addProcessor(
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: I'm having a hard time finding the declaration of this method. I can't find it in your code nor in the existing master branch. Can you link to it?

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 2]
```diff
@@ -2350,13 +2354,12 @@ class SparkContext(config: SparkConf) extends Logging {
     try {
       val listenerClassNames: Seq[String] =
         conf.get("spark.extraListeners", "").split(',').map(_.trim).filter(_ != "")
-      for (className <- listenerClassNames) {
-        // Use reflection to find the right constructor
+      val extraListeners = listenerClassNames.map{ className =>
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: `.map { className =>`

You have a lot of style issues in your code - indentation, spacing, etc. Please read the style section in http://spark.apache.org/contributing.html and try to follow it.

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 3]
```diff
@@ -2379,8 +2382,13 @@ class SparkContext(config: SparkConf) extends Logging {
                 " parameter from breaking Spark's ability to find a valid constructor.")
           }
         }
-        listenerBus.addListener(listener)
-        logInfo(s"Registered listener $className")
+        logInfo(s"listener $className created")
+        listener
+      }
+      if (extraListeners.nonEmpty) {
+        val group = new FixGroupOfListener(extraListeners, "extraListeners")
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: `FixGroupOfListener` is a bad class name. I'm not even sure what it's supposed to be, but the closest I can think of is `ListenerGroup`.

But perhaps this shouldn't be exposed at all. If you add a queue name to the listener registration method, you can hide this from callers altogether. That is, if I understood what this class is in the first place.

Then you wouldn't need `addIsolatedListener` either.

File: core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala [hunk 4]
```diff
@@ -227,6 +169,7 @@ private[spark] class EventLoggingListener(
    * ".inprogress" suffix.
    */
   def stop(): Unit = {
+    flush()
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala]: Shouldn't be necessary (`close()` does it).

File: core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala [hunk 5]
```diff
@@ -131,94 +132,35 @@ private[spark] class EventLoggingListener(
...
-
-  // No-op because logging every update would be overkill
-  override def onBlockUpdated(event: SparkListenerBlockUpdated): Unit = {}
-
-  // No-op because logging every update would be overkill
-  override def onExecutorMetricsUpdate(event: SparkListenerExecutorMetricsUpdate): Unit = { }
-
-  override def onOtherEvent(event: SparkListenerEvent): Unit = {
+  def log(event: SparkListenerEvent): Unit = {
     if (event.logEvent) {
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala]: Since you're adding an event filter, you could perform this check there...

File: core/src/main/scala/org/apache/spark/scheduler/bus/ListenerBusQueueImpl.scala [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+package org.apache.spark.scheduler.bus
+
+import com.codahale.metrics.Timer
+import scala.reflect.ClassTag
+
+import org.apache.spark.scheduler.{SparkListenerEvent, SparkListenerEventDispatcher, SparkListenerInterface}
+import org.apache.spark.scheduler.bus.ListenerBusQueue.{FixGroupOfListener, ModifiableGroupOfListener}
+
+// For generic message processor (like event logging)
+private[scheduler] class ProcessorListenerBusQueue(
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/scheduler/bus/ListenerBusQueueImpl.scala]: First, the name of this file is weird. But more importantly, why are these classes even necessary?

Why can't you have a single queue implementation that manages a group of listeners? Whether the group has a single listener or multiple shouldn't matter - the implementation can be the same.

**Review by vanzin — COMMENTED** [review]
I started reviewing, but again, I noticed the same thing I commented on before. This is way over-engineered. You can do this in a much, much simpler way. There's no need to create all the different abstractions you're adding - the current listener abstraction is enough to achieve what is being proposed here. 

All you need is to add a "queue name" parameter to the `addListener` method, and potentially an "event filter" parameter. Everything else is hidden in the listener implementation, and doesn't need to be exposed to any calling code.

File: core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala [hunk 7]
```diff
@@ -139,30 +148,60 @@ private[spark] class LiveListenerBus(conf: SparkConf) extends SparkListenerBus {
...
+    startStopAddRemoveLock.lock()
+    if (!started) {
+      sparkContext = sc
+      metricsSystem = ms
+      defaultListenerQueue.start(sc, ms)
+      otherListenerQueues.foreach(_.start(sc, ms))
+      started = true
     } else {
-      onDropEvent(event)
+      throw new IllegalStateException("LiveListener bus already started!")
```
Comment (abellina) [inline core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala]: definitely want to unlock before this.

File: core/src/main/scala/org/apache/spark/scheduler/SparkListenerBus.scala [hunk 8]
```diff
@@ -27,7 +27,12 @@ private[spark] trait SparkListenerBus
 
   protected override def doPostEvent(
       listener: SparkListenerInterface,
-      event: SparkListenerEvent): Unit = {
+      event: SparkListenerEvent): Unit = SparkListenerEventDispatcher.dispatch(listener, event)
```
Comment (abellina) [inline core/src/main/scala/org/apache/spark/scheduler/SparkListenerBus.scala]: why is this change necessary?

**Review by abellina — COMMENTED** [review]
Hi just some comments as I try to understand the code.

File: core/src/main/scala/org/apache/spark/scheduler/bus/ListenerBusQueueImpl.scala (hunk 6 above)
Comment (bOOm-X) [inline core/src/main/scala/org/apache/spark/scheduler/bus/ListenerBusQueueImpl.scala]: For the name of the file, I can change it ! Do you have a better name ? I can even put the content of the file (the 2 concrete implementations)  in the BusQueue.scala file.

I refactored a bit this file. Now I have only 2 implementations: 
 - `ProcessorBusQueue`: This is the implementation for generic processor (In which we do not do the dispatch by  event type)
-  `ListenerBusQueue`: This is the implementation for listener (with the dispatch by event type)

**Comment by bOOm-X** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/SparkContext.scala (hunk 3 above)
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: > First, I think that modifying the existing addListener method is a bad idea. It will impact a lot the code.

That's why overloaded methods exist.

> But I have 2 kind of group of listeners

I don't think there's really a distinction between the two types of groups you mention. The "UI group" is just a modifiable group that you don't modify after it's been created.

File: core/src/main/scala/org/apache/spark/scheduler/bus/ListenerBusQueueImpl.scala (hunk 6 above)
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/scheduler/bus/ListenerBusQueueImpl.scala]: I'm still confused about why you need 2 implementations. Why doesn't `ListenerBusQueue` work for everybody? And why shouldn't it?

I need more time to actually grok all this code, but like Wenchen suggested before, this is a big change and it would benefit from a more detailed explanation of exactly how you're organizing the hierarchy of listener, groups, etc. Your PR description only explains which queues you created, but not any of the changes that were needed to achieve that.

If it makes it easier, you can create a `README.md` file with a longer explanation for how things are organized. (for example, check `ommon/network-common/src/main/java/org/apache/spark/network/crypto` where I added a README to explain details of what that whole body of code is doing).

**Comment by bOOm-X** [comment]
@vanzin I simplify a lot the code. There are now only one implementation for the queue and for the group of listeners. I removed the extra trait in the listener hierarchy too.
Can you take a look ?

File: core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala [hunk 9]
```diff
@@ -39,98 +40,114 @@ import org.apache.spark.util.Utils
...
+    }
+    startStopAddRemoveLock.unlock()
+  }
 
-  // A counter that represents the number of events produced and consumed in the queue
-  private val eventLock = new Semaphore(0)
+   /**
+    * Add a generic listener to an isolated pool.
+    */
+  def addProcessor(processor: SparkListenerEvent => Unit,
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala]: what's the difference between `addProcessor` and `addListener`?

**Comment by cloud-fan** [comment]
The PR description is good for explaining the new behavior, but can you say more about the implementation?

IMO, we just need to duplicate the event queue for each important listener like event logging listener, and non-important listeners can share one event queue like the current behavior. Then each event queue is processed by an individual thread.

File: core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala (hunk 9 above)
Comment (bOOm-X) [inline core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala]: with addProcessor, you do not have to provide a SparkListenerInterface (So with a method per message type), but just a generic function which handle  SparkListenerEvent  (the super type of each event type). So when you do a generic processing (see EventLoggingListener for example)  it is very convenient, and cherry on the top you avoid the horrible and costly dispatch function , which in this case (generic processing) is a burden

**Comment by bOOm-X** [comment]
@cloud-fan I will update the PR description with some details on the implementation

File: core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala (hunk 9 above)
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala]: shall we do it in a separated PR?

Comment (bOOm-X) [inline core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala]: It is just a small technical refactoring which came for almost free with the new qeue object. It is also very convenient to be able to handle the asynchronous LiveListenerBus in the test.  I think that we can keep it in this PR.

**Comment by cloud-fan** [comment]
I think `BusQueue` is a good abstraction, but we can still simplify other parts. My proposal:
```
// Do we need a better name like ListenesGroup? it's very similar to the previous LiveListenerBus.
class BusQueue extends SparkListenerBus {
  val eventQueue = ...
}

class LiveListenerBus {
  val listenersGroups: Map[String, BusQueue] = ...

  def addListener(listener, groupName = "default") {
    val group = listenerGroups.getOrUpdate(groupName, new BusQueue ...)
    group.addListener(listener)
  }

  def post(event) {
    listenersGroups.foreach { case (_, group) =>
      group.post(event)
    }
  }
}

```

**Comment by vanzin** [comment]
Agree that it still seems like there's too many moving parts here. I don't see a whole lot of difference between `BusQueue` and `GroupOfListener` (which is a weird name, btw). A queue can have a bunch of listeners and a `LiveListenerBus` can have a bunch of queues. I don't see the need for anything more complicated than that.

`WithListenerBus` is another thing I don't understand; aside from the really weird name, it seems to be mostly the same thing as `ListenerBus`.

Also to reinforce a previous comment, your code has a ton of style issues. It doesn't matter that checkstyle doesn't complain about them; you still have to follow the code convention of the project or we'll be forever pointing out style issues in your code.

**Comment by bOOm-X** [comment]
@cloud-fan 
In my opinion, having the queues indexed by a string and reflecting that in the API is a bit too error prone in this case. What you want is to be in an isolated queue. With this kind of API it is easy to have an other listener added to the same queue because of a conflicting label. In the contrary, it is quite easy to not had 2 depending listeners not in the same queue because of case, ..., of the label. With a map of queues, it enforces the mutability of the queue, which is not really necessary here. And I think that it is a good idea to enforce the fact that depending listeners should be put in the same queue, it increases the readability and the clarity of the API.

**Comment by bOOm-X** [comment]
@vanzin 
`GroupOfListener` is just a set of listeners, which handle all the metric stuff. it allows to decouple this metric aspect from the `ListernerBus` and the `LiveListenerBus`. And It allows to enforce the dependency model between listeners (one message processed sequentially). I can change its name if you want. 
The `BusQueue` just handle the queuing / dequeueing generic processing, which includes the consumer thread start an stop and the queuing strategy (drop if full).

`WithListenerBus` is indeed just the declaration of the method implemented in `ListenerBus`. This declaration is shared between the `LiveListenerBus` and the `ReplayListenerBus`, and the UI listeners are added to them through it. The `ListenerBus` still contains a simple synchronous implementation of them used by the `ReplayListenerBus`. The `LiveListenerBus` has its own based on multiple queues for some kind of set of listeners. I can rename this interface if you want.

I will do a pass to try to fix the code style issues.

**Comment by vanzin** [comment]
I really dislike `WithListenerBus` - both as a name and as a concept. There's already a `ListenerBus` trait; if it's not enough or is broken in some way, it should be fixed, instead of being patched by introducing yet more complexity in the hierarchy.

I think part of the confusion here is that the current code is trying to both refactor the `ListenerBus` hierarchy and add the concept of queues are the same time. From my point of view they're different things and could be done separately. For example you could add queues to `LiveListenerBus` only, which is really the only place that matters in the end. Maybe it won't be optimal in its first iteration, but it would be a much easier change to review.

I don't doubt that there's benefit in taking a holistic look into this part of the class hierarchy; but it would be good to do that separately, both so that we can clearly see that the proposed hierarchy makes sense, and so that it's easier to review things. It's easier to wrap your head around the code if it's focused on one problem instead of two.

**Comment by vanzin** [comment]
@bOOm-X 

I pushed some code to my repo: https://github.com/vanzin/spark/tree/SPARK-18838

Which is an attempt to do things the way I've been trying to explain. It tries to keep changes as local as possible to `LiveListenerBus`, creating a few types that implement the grouping and asynchronous behavior. You could do filtering by extending the new `AsyncListener`, for example, and adding it to the live listener bus.

It's just a p.o.c. so I cut a few corners (like metrics), and I only ran `SparkListenerSuite`, but I'm just trying to show a different approach that leaves the `ListenerBus` hierarchy mostly the same as now.

**Comment by bOOm-X** [comment]
@vanzin I pushed some comments on your code. I think that trying to keep the exact same class hierarchy leads to a very complex code, with many drawbacks.

**Comment by vanzin** [comment]
You commented on my code, not on the idea. My code was hacked together quickly, it can be cleaned up a lot. Your comments don't disprove that separating the refactoring of the listener bus hierarchy from the introduction of queues is impossible or undesirable.
